File: src/sphere.ts
import { HitRecord, Hittable } from "./hittable";
import { Interval } from "./interval";
import { Ray } from "./ray";
import { Vec3 } from "./vec3";

export class Sphere extends Hittable {
  constructor(
    public center: Vec3,
    public radius: number,
  ) {
    super();
  }

  hit(r: Ray, rayT: Interval, rec: HitRecord): [boolean, HitRecord] {
    const oc = this.center.sub(r.origin);
    const a = r.direction.lengthSquared;
    const h = Vec3.dot(r.direction, oc);
    const c = oc.lengthSquared - this.radius * this.radius;
    const discriminant = h * h - a * c;

    if (discriminant < 0) {
      return [false, rec];
    }

    let root = (h - Math.sqrt(discriminant)) / a;

    if (!rayT.surrounds(root)) {
      root = (h + Math.sqrt(discriminant)) / a;

      if (!rayT.surrounds(root)) {
        return [false, rec] as const;
      }
    }

    rec.t = root;
    rec.p = r.at(rec.t);
    let outwardNormal = rec.p.sub(this.center).div(this.radius);
    rec.setFaceNormal(r, outwardNormal);

    return [true, rec] as const;
  }
}

File: raytracer/sphere.go
package raytracer

import (
	"math"
	"raytracer/math3"
)

type Sphere struct {
	Center   math3.Vec3
	Radius   float64
	Material Material
}

func (s Sphere) Hit(ray math3.Ray, rayT Interval) (HitRecord, bool) {
	oc := s.Center.Sub(ray.Origin)
	a := ray.Direction.LengthSquared()
	h := math3.Dot(ray.Direction, oc)
	c := oc.LengthSquared() - s.Radius*s.Radius
	disc := h*h - a*c
	if disc < 0 {
		return HitRecord{}, false
	}
	sqrtd := math.Sqrt(disc)
	root := (h - sqrtd) / a
	if !rayT.Surrounds(root) {
		root = (h + sqrtd) / a
		if !rayT.Surrounds(root) {
			return HitRecord{}, false
		}
	}
	rec := HitRecord{}
	rec.T = root
	rec.P = ray.At(rec.T)
	outwardNormal := rec.P.Sub(s.Center).Div(s.Radius)
	rec.SetFaceNormal(ray, outwardNormal)
	rec.Material = s.Material
	return rec, true
}
